### DISCUSSION TITLE
More volumes are causing more IO on RKE2 server nodes

### DISCUSSION THREAD

**Discussion by serhiynovos** [longhorn/longhorn#8543] [General]
I'm managing an RKE2 cluster with Longhorn 1.6.1 in an on-prem setup. Yesterday, I installed MinIO using 9 volumes, each configured without replication. I've noticed an increase in the write rate. Although the increase is modest compared to the previous version 1.5.4, our future plans involve deploying approximately 300 additional volumes, most with a replication factor of 3. I'm concerned about the potential I/O load on my server/etcd nodes. Could Longhorn handle this increased load without prematurely wearing out the SSDs on server nodes? Is this behavior normal, or should I be looking into potential issues?

This is my current state with replicas
<img width="904" alt="image" src="https://github.com/longhorn/longhorn/assets/9201885/7f26f5e9-d214-44f4-9faf-904ae209bc5b">

And how IO rate changing after installing MINIO with total 9 replicas
![image](https://github.com/longhorn/longhorn/assets/9201885/d481399b-1dbc-4bd3-89c2-f66e845c2b82)

**Comment by serhiynovos** [comment]
Can anyone confirm that it's normal behaviour or I have to dig deeper to see what is causing higher load. Thanks

**Comment by derekbit** [comment]
Hello @serhiynovos,

Do you mean the number of write operations has increased after installing MinIO? If that's the case, it's likely due to MinIO, and you can investigate why MinIO is writing data to the volumes more frequently. This increase is independent of the type of storage used.
An increase in write operations could shorten the lifespan of the SSDs. However, since object storage systems typically write data using an append method, this approach can mitigate the wearout effect.

**Reply by serhiynovos** [reply]
Hi @derekbit ,
Thank you for your response. More writes happens on server + etcd nodes not where particular volume is located. Based on screenshot I provided after adding 9 more no replica volumes created by minio operator I have about 70 more kb/s write rate on server + etcd node. For now it's not clear if it's caused by minio or just by adding extra 9 volumes

As I'm relatively new to Kubernetes, I don't know if it's normal behavior or not. Now we are working on our own Kubernetes cluster where will will provide SASS for our clients and as on current state our apps don't support soft multitenancy we will have to run separate application stack for each customer which can lead creating a big amount of additional volumes for now I see on this particular cluster can be from 200 to 300 additional volumes and it's not clear what impact it will cause on server + etcd nodes

**Reply by derekbit** [reply]
@serhiynovos Thanks for the update. I thought you meant the longhorn volume. After creating volumes, Longhorn will periodically collect information such as disk info and so on. This information will be updated to volume or other resources, so an increase in the write rate is expected.

**Reply by serhiynovos** [reply]
@derekbit so you think it should not create any issues to have 300 extra volumes where each of them may have 3 replicas? Or there is some limits to volumes count ? I don't think so it will be high load workloads, just bunch of them.

**Reply by derekbit** [reply]
I think we can benchmark the write rates in our lab environment first and compare them with your numbers.

cc @PhanLe1010 @shuo-wu

**Reply by serhiynovos** [reply]
@derekbit Thank you. It will help so much. 
Maybe it will require to modify some services which now needs volumes to store static files as images, videos. So instead local volume it can use minio deployed in the same cluster where I'll have only a few (about 9) but big volumes and each client will have  own bucket which should reduce amount of volumes and reduce load to the etcd.

**Reply by PhanLe1010** [reply]
> I think we can benchmark the write rates in our lab environment first and compare them with your numbers.

Eric has a good calculation below at https://github.com/longhorn/longhorn/discussions/8543#discussioncomment-9446713 so we can continue over there

**Comment by ejweber** [comment]
We have https://longhorn.io/kb/kubernetes-resource-revision-frequency-expectations/ describing the expectations for API server PUT operations in a stable cluster. In that investigation, we focused on update frequency and not on I/O throughput, but it may help explain the behavior your are observing.

There is an improvement planned in https://github.com/longhorn/longhorn/issues/8076 that should help to lower the update frequency for engine and volume objects.

If you have Prometheus active in the cluster, it may help to run some queries like we did in https://github.com/longhorn/longhorn/issues/8114#issuecomment-1988842948 to see if they are in line with the "worst case" scenario for a stable cluster in the knowledgebase (i.e. 12 PUTs per minute per engine resource and 12 PUTs per minute per volume resource).

**Reply by ejweber** [reply]
As part of a previous investigation, I used `etcdctl` to investigate stored Longhorn objects.

For example, here is a Longhorn volume:

```
root@eweber-v126-master-7ccf4f43-spxmg:~# etcdctl --cacert=/var/lib/rancher/k3s/server/tls/etcd/server-ca.crt --cert=/var/lib/rancher/k3s/server/tls/etcd/client.crt --key=/var/lib/rancher/k3s/server/tls/etcd/client.key get "/registry/longhorn.io/volumes/longhorn-system/pvc-9dcfec90-ad55-469e-b0a5-473dd5391ba0" -w json | jq
{
  "header": {
    "cluster_id": 10586671132190937000,
    "member_id": 8930079509784360000,
    "revision": 47949401,
    "raft_term": 4
  },
  "kvs": [
    {
      "key": "L3JlZ2lzdHJ5L2xvbmdob3JuLmlvL3ZvbHVtZXMvbG9uZ2hvcm4tc3lzdGVtL3B2Yy05ZGNmZWM5MC1hZDU1LTQ2OWUtYjBhNS00NzNkZDUzOTFiYTA=",
      "create_revision": 47947680,
      "mod_revision": 47949373,
      "version": 48,
      "value": "eyJhcGlWZXJzaW9uIjoibG9uZ2hvcm4uaW8vdjFiZXRhMiIsImtpbmQiOiJWb2x1bWUiLCJtZXRhZGF0YSI6eyJjcmVhdGlvblRpbWVzdGFtcCI6IjIwMjQtMDItMjNUMTU6MzM6NDhaIiwiZmluYWxpemVycyI6WyJsb25naG9ybi5pbyJdLCJnZW5lcmF0aW9uIjoyLCJsYWJlbHMiOnsibG9uZ2hvcm52b2x1bWUiOiJwdmMtOWRjZmVjOTAtYWQ1NS00NjllLWIwYTUtNDczZGQ1MzkxYmEwIiwicmVjdXJyaW5nLWpvYi1ncm91cC5sb25naG9ybi5pby9kZWZhdWx0IjoiZW5hYmxlZCIsInNldHRpbmcubG9uZ2hvcm4uaW8vcmVtb3ZlLXNuYXBzaG90cy1kdXJpbmctZmlsZXN5c3RlbS10cmltIjoiaWdub3JlZCIsInNldHRpbmcubG9uZ2hvcm4uaW8vcmVwbGljYS1hdXRvLWJhbGFuY2UiOiJpZ25vcmVkIiwic2V0dGluZy5sb25naG9ybi5pby9zbmFwc2hvdC1kYXRhLWludGVncml0eSI6Imlnbm9yZWQifSwibWFuYWdlZEZpZWxkcyI6W3siYXBpVmVyc2lvbiI6Imxvbmdob3JuLmlvL3YxYmV0YTIiLCJmaWVsZHNUeXBlIjoiRmllbGRzVjEiLCJmaWVsZHNWMSI6eyJmOm1ldGFkYXRhIjp7ImY6bGFiZWxzIjp7Ii4iOnt9LCJmOnJlY3VycmluZy1qb2ItZ3JvdXAubG9uZ2hvcm4uaW8vZGVmYXVsdCI6e319fSwiZjpzcGVjIjp7Ii4iOnt9LCJmOlN0YW5kYnkiOnt9LCJmOmFjY2Vzc01vZGUiOnt9LCJmOmJhY2tlbmRTdG9yZURyaXZlciI6e30sImY6YmFja2luZ0ltYWdlIjp7fSwiZjpiYWNrdXBDb21wcmVzc2lvbk1ldGhvZCI6e30sImY6ZGF0YUVuZ2luZSI6e30sImY6ZGF0YUxvY2FsaXR5Ijp7fSwiZjpkYXRhU291cmNlIjp7fSwiZjpkaXNhYmxlRnJvbnRlbmQiOnt9LCJmOmVuY3J5cHRlZCI6e30sImY6ZW5naW5lSW1hZ2UiOnt9LCJmOmZyb21CYWNrdXAiOnt9LCJmOmZyb250ZW5kIjp7fSwiZjppbWFnZSI6e30sImY6bGFzdEF0dGFjaGVkQnkiOnt9LCJmOm1pZ3JhdGFibGUiOnt9LCJmOm1pZ3JhdGlvbk5vZGVJRCI6e30sImY6bm9kZUlEIjp7fSwiZjpudW1iZXJPZlJlcGxpY2FzIjp7fSwiZjpvZmZsaW5lUmVwbGljYVJlYnVpbGRpbmciOnt9LCJmOnJlcGxpY2FBdXRvQmFsYW5jZSI6e30sImY6cmVwbGljYURpc2tTb2Z0QW50aUFmZmluaXR5Ijp7fSwiZjpyZXBsaWNhU29mdEFudGlBZmZpbml0eSI6e30sImY6cmVwbGljYVpvbmVTb2Z0QW50aUFmZmluaXR5Ijp7fSwiZjpyZXN0b3JlVm9sdW1lUmVjdXJyaW5nSm9iIjp7fSwiZjpyZXZpc2lvbkNvdW50ZXJEaXNhYmxlZCI6e30sImY6c2l6ZSI6e30sImY6c25hcHNob3REYXRhSW50ZWdyaXR5Ijp7fSwiZjpzbmFwc2hvdE1heENvdW50Ijp7fSwiZjpzbmFwc2hvdE1heFNpemUiOnt9LCJmOnN0YWxlUmVwbGljYVRpbWVvdXQiOnt9LCJmOnVubWFwTWFya1NuYXBDaGFpblJlbW92ZWQiOnt9fX0sIm1hbmFnZXIiOiJsb25naG9ybi1tYW5hZ2VyIiwib3BlcmF0aW9uIjoiVXBkYXRlIiwidGltZSI6IjIwMjQtMDItMjNUMTU6MzM6NTNaIn0seyJhcGlWZXJzaW9uIjoibG9uZ2hvcm4uaW8vdjFiZXRhMiIsImZpZWxkc1R5cGUiOiJGaWVsZHNWMSIsImZpZWxkc1YxIjp7ImY6c3RhdHVzIjp7Ii4iOnt9LCJmOmFjdHVhbFNpemUiOnt9LCJmOmNsb25lU3RhdHVzIjp7Ii4iOnt9LCJmOnNuYXBzaG90Ijp7fSwiZjpzb3VyY2VWb2x1bWUiOnt9LCJmOnN0YXRlIjp7fX0sImY6Y29uZGl0aW9ucyI6e30sImY6Y3VycmVudEltYWdlIjp7fSwiZjpjdXJyZW50TWlncmF0aW9uTm9kZUlEIjp7fSwiZjpjdXJyZW50Tm9kZUlEIjp7fSwiZjpleHBhbnNpb25SZXF1aXJlZCI6e30sImY6ZnJvbnRlbmREaXNhYmxlZCI6e30sImY6aXNTdGFuZGJ5Ijp7fSwiZjprdWJlcm5ldGVzU3RhdHVzIjp7Ii4iOnt9LCJmOmxhc3RQVkNSZWZBdCI6e30sImY6bGFzdFBvZFJlZkF0Ijp7fSwiZjpuYW1lc3BhY2UiOnt9LCJmOnB2TmFtZSI6e30sImY6cHZTdGF0dXMiOnt9LCJmOnB2Y05hbWUiOnt9LCJmOndvcmtsb2Fkc1N0YXR1cyI6e319LCJmOmxhc3RCYWNrdXAiOnt9LCJmOmxhc3RCYWNrdXBBdCI6e30sImY6bGFzdERlZ3JhZGVkQXQiOnt9LCJmOm9mZmxpbmVSZXBsaWNhUmVidWlsZGluZ1JlcXVpcmVkIjp7fSwiZjpvd25lcklEIjp7fSwiZjpwZW5kaW5nTm9kZUlEIjp7fSwiZjpyZW1vdW50UmVxdWVzdGVkQXQiOnt9LCJmOnJlc3RvcmVJbml0aWF0ZWQiOnt9LCJmOnJlc3RvcmVSZXF1aXJlZCI6e30sImY6cm9idXN0bmVzcyI6e30sImY6c2hhcmVFbmRwb2ludCI6e30sImY6c2hhcmVTdGF0ZSI6e30sImY6c3RhdGUiOnt9fX0sIm1hbmFnZXIiOiJsb25naG9ybi1tYW5hZ2VyIiwib3BlcmF0aW9uIjoiVXBkYXRlIiwic3VicmVzb3VyY2UiOiJzdGF0dXMiLCJ0aW1lIjoiMjAyNC0wMi0yM1QxNTozNzoyN1oifV0sIm5hbWUiOiJwdmMtOWRjZmVjOTAtYWQ1NS00NjllLWIwYTUtNDczZGQ1MzkxYmEwIiwibmFtZXNwYWNlIjoibG9uZ2hvcm4tc3lzdGVtIiwidWlkIjoiMTU1OTdmZDctOTA0NC00ODVkLTgzY2MtYWI4NmYzMDdkNzNlIn0sInNwZWMiOnsiU3RhbmRieSI6ZmFsc2UsImFjY2Vzc01vZGUiOiJyd28iLCJiYWNrZW5kU3RvcmVEcml2ZXIiOiIiLCJiYWNraW5nSW1hZ2UiOiIiLCJiYWNrdXBDb21wcmVzc2lvbk1ldGhvZCI6Imx6NCIsImRhdGFFbmdpbmUiOiJ2MSIsImRhdGFMb2NhbGl0eSI6ImRpc2FibGVkIiwiZGF0YVNvdXJjZSI6IiIsImRpc2FibGVGcm9udGVuZCI6ZmFsc2UsImRpc2tTZWxlY3RvciI6W10sImVuY3J5cHRlZCI6ZmFsc2UsImVuZ2luZUltYWdlIjoiIiwiZnJvbUJhY2t1cCI6IiIsImZyb250ZW5kIjoiYmxvY2tkZXYiLCJpbWFnZSI6Imxvbmdob3JuaW8vbG9uZ2hvcm4tZW5naW5lOnYxLjYuMCIsImxhc3RBdHRhY2hlZEJ5IjoiIiwibWlncmF0YWJsZSI6ZmFsc2UsIm1pZ3JhdGlvbk5vZGVJRCI6IiIsIm5vZGVJRCI6ImV3ZWJlci12MTI2LXdvcmtlci05YzE0NTFiNC1rZ3hkcSIsIm5vZGVTZWxlY3RvciI6W10sIm51bWJlck9mUmVwbGljYXMiOjMsIm9mZmxpbmVSZXBsaWNhUmVidWlsZGluZyI6ImRpc2FibGVkIiwicmVwbGljYUF1dG9CYWxhbmNlIjoiaWdub3JlZCIsInJlcGxpY2FEaXNrU29mdEFudGlBZmZpbml0eSI6Imlnbm9yZWQiLCJyZXBsaWNhU29mdEFudGlBZmZpbml0eSI6Imlnbm9yZWQiLCJyZXBsaWNhWm9uZVNvZnRBbnRpQWZmaW5pdHkiOiJpZ25vcmVkIiwicmVzdG9yZVZvbHVtZVJlY3VycmluZ0pvYiI6Imlnbm9yZWQiLCJyZXZpc2lvbkNvdW50ZXJEaXNhYmxlZCI6ZmFsc2UsInNpemUiOiIyMTQ3NDgzNjQ4Iiwic25hcHNob3REYXRhSW50ZWdyaXR5IjoiaWdub3JlZCIsInNuYXBzaG90TWF4Q291bnQiOjI1MCwic25hcHNob3RNYXhTaXplIjoiMCIsInN0YWxlUmVwbGljYVRpbWVvdXQiOjMwLCJ1bm1hcE1hcmtTbmFwQ2hhaW5SZW1vdmVkIjoiaWdub3JlZCJ9LCJzdGF0dXMiOnsiYWN0dWFsU2l6ZSI6MTAyNjc4NTI4LCJjbG9uZVN0YXR1cyI6eyJzbmFwc2hvdCI6IiIsInNvdXJjZVZvbHVtZSI6IiIsInN0YXRlIjoiIn0sImNvbmRpdGlvbnMiOlt7Imxhc3RQcm9iZVRpbWUiOiIiLCJsYXN0VHJhbnNpdGlvblRpbWUiOiIyMDI0LTAyLTIzVDE1OjMzOjQ5WiIsIm1lc3NhZ2UiOiIiLCJyZWFzb24iOiIiLCJzdGF0dXMiOiJGYWxzZSIsInR5cGUiOiJXYWl0Rm9yQmFja2luZ0ltYWdlIn0seyJsYXN0UHJvYmVUaW1lIjoiIiwibGFzdFRyYW5zaXRpb25UaW1lIjoiMjAyNC0wMi0yM1QxNTozMzo0OVoiLCJtZXNzYWdlIjoiIiwicmVhc29uIjoiIiwic3RhdHVzIjoiRmFsc2UiLCJ0eXBlIjoiVG9vTWFueVNuYXBzaG90cyJ9LHsibGFzdFByb2JlVGltZSI6IiIsImxhc3RUcmFuc2l0aW9uVGltZSI6IjIwMjQtMDItMjNUMTU6MzM6NDlaIiwibWVzc2FnZSI6IiIsInJlYXNvbiI6IiIsInN0YXR1cyI6IlRydWUiLCJ0eXBlIjoiU2NoZWR1bGVkIn0seyJsYXN0UHJvYmVUaW1lIjoiIiwibGFzdFRyYW5zaXRpb25UaW1lIjoiMjAyNC0wMi0yM1QxNTozMzo0OVoiLCJtZXNzYWdlIjoiIiwicmVhc29uIjoiIiwic3RhdHVzIjoiRmFsc2UiLCJ0eXBlIjoiUmVzdG9yZSJ9XSwiY3VycmVudEltYWdlIjoibG9uZ2hvcm5pby9sb25naG9ybi1lbmdpbmU6djEuNi4wIiwiY3VycmVudE1pZ3JhdGlvbk5vZGVJRCI6IiIsImN1cnJlbnROb2RlSUQiOiJld2ViZXItdjEyNi13b3JrZXItOWMxNDUxYjQta2d4ZHEiLCJleHBhbnNpb25SZXF1aXJlZCI6ZmFsc2UsImZyb250ZW5kRGlzYWJsZWQiOmZhbHNlLCJpc1N0YW5kYnkiOmZhbHNlLCJrdWJlcm5ldGVzU3RhdHVzIjp7Imxhc3RQVkNSZWZBdCI6IiIsImxhc3RQb2RSZWZBdCI6IiIsIm5hbWVzcGFjZSI6ImRlZmF1bHQiLCJwdk5hbWUiOiJwdmMtOWRjZmVjOTAtYWQ1NS00NjllLWIwYTUtNDczZGQ1MzkxYmEwIiwicHZTdGF0dXMiOiJCb3VuZCIsInB2Y05hbWUiOiJsb25naG9ybi12b2x2LXB2YyIsIndvcmtsb2Fkc1N0YXR1cyI6W3sicG9kTmFtZSI6InZvbHVtZS10ZXN0IiwicG9kU3RhdHVzIjoiUnVubmluZyIsIndvcmtsb2FkTmFtZSI6IiIsIndvcmtsb2FkVHlwZSI6IiJ9XX0sImxhc3RCYWNrdXAiOiIiLCJsYXN0QmFja3VwQXQiOiIiLCJsYXN0RGVncmFkZWRBdCI6IiIsIm9mZmxpbmVSZXBsaWNhUmVidWlsZGluZ1JlcXVpcmVkIjpmYWxzZSwib3duZXJJRCI6ImV3ZWJlci12MTI2LXdvcmtlci05YzE0NTFiNC1rZ3hkcSIsInBlbmRpbmdOb2RlSUQiOiIiLCJyZW1vdW50UmVxdWVzdGVkQXQiOiIiLCJyZXN0b3JlSW5pdGlhdGVkIjpmYWxzZSwicmVzdG9yZVJlcXVpcmVkIjpmYWxzZSwicm9idXN0bmVzcyI6ImhlYWx0aHkiLCJzaGFyZUVuZHBvaW50IjoiIiwic2hhcmVTdGF0ZSI6IiIsInN0YXRlIjoiYXR0YWNoZWQifX0K"
    }
  ],
  "count": 1
}
```

According to `wc`, the size of the `value` field in bytes is:

```
eweber@laptop:/tmp> wc -c volume_value.txt 
6393 volume_value.txt # Approximately 6 kiB.
```

So if your API server PUT rate is the worst case from above, it could lead to:

```
(9 volumes)(12 PUTs / minute / volume)(6393 B / PUT) / [(1024 kiB / B)(60 s / min)] = 11 kiB/s
```

Probably the corresponding engines will contribute a similar rate of writes.

This is all back of the napkin math. Hopefully we can check etcd write rates in the lab.

**Reply by PhanLe1010** [reply]
Btw, to enable monitoring:
1. If using Rancher https://ranchermanager.docs.rancher.com/integrations-in-rancher/monitoring-and-alerting
2. Otherwise, this is the upstream Prometheus monitoring stack chart https://github.com/prometheus-community/helm-charts/blob/main/charts/kube-prometheus-stack/README.md

**Reply by serhiynovos** [reply]
Yes I have installed Prometheus on my cluster and now I don't see anything what was doing something strange. I'm aware about issue in 1.5.4 regarding high PUT rate. I also faced it before https://github.com/rancher/rke2/discussions/5680 :) So it's not related to it. Currently minio is uninstalled and I don't see any difference in these 2 metrics. 
<img width="1727" alt="Знімок екрана 2024-05-15 о 21 01 29" src="https://github.com/longhorn/longhorn/assets/9201885/f986ccc5-8482-4855-a1e2-2a3c52771c94">
<img width="1728" alt="Знімок екрана 2024-05-15 о 21 02 42" src="https://github.com/longhorn/longhorn/assets/9201885/432c59b6-87c0-42d4-8bf5-8f01341f406c">
<img width="1728" alt="Знімок екрана 2024-05-15 о 21 03 39" src="https://github.com/longhorn/longhorn/assets/9201885/a9afc258-39f9-4474-afdc-aef37dc728e0">

@ejweber  I think you are right. It can put additional 11 kb/s write rate to ETCD + pods statuses as well can be updated on etcd side so it can add few more extra kb/s for write rate. On the chart I noticed it had about 40kb/s increase. 

`(9 volumes)(12 PUTs / minute / volume)(6393 B / PUT) / [(1024 kiB / B)(60 s / min)] = 11 kiB/s` 

If every volume has 3 replicas this rate I can multiply by 3. Right ?

**Reply by ejweber** [reply]
> If every volume has 3 replicas this rate I can multiply by 3. Right ?

Probably no. In the pending https://github.com/longhorn/longhorn/issues/8076 we focus on engine and replica resources because they have a `size` field that Longhorn updates regularly, even when nothing else "interesting" is going on. The replica object doesn't have a similar field, so it shouldn't experience a high rate of PUT requests.

Unfortunately, I cannot see the full key/legend in the screenshots you provided. In the last one, I only see LIST/WATCH/DELETE in the key/legend, but I think only CREATE/PUT/etc. will lead to the extra writing.

- Which API/verb combos are associated with the high lines on the graph?
- At what point on the graph did you add volumes?

**Reply by serhiynovos** [reply]
`At what point on the graph did you add volumes?`
I installed Prometheus right after I installed minio with 9 volumes to check if nothing is not doing a lot of requests to API server. About in a middle of graph I uninstalled minio and removed volumes and as I see nothing has changed on graph. 

`Which API/verb combos are associated with the high lines on the graph?`
Top 3. GET. 4th PUT from. All 4 lines from the same group "coordination.k8s.io"
<img width="1717" alt="Знімок екрана 2024-05-15 о 21 40 58" src="https://github.com/longhorn/longhorn/assets/9201885/48c7a8d2-f80a-41e7-ab47-844901309718">
<img width="1714" alt="Знімок екрана 2024-05-15 о 21 41 14" src="https://github.com/longhorn/longhorn/assets/9201885/37702f10-43e2-41e8-90d3-4112393278f9">

`Probably no`. 
Which means if I have 300 volumes they have to cause about 360 kb/s which is totally fine for this amount of volumes. But here is another question for this kind of scale. I know longhorn is using iSCSI. Can it be a bottleneck as I know it's using TCP protocol and even pod and volume located on the same node it's not directly mounted to the pod but all is going through TCP connection?

**Reply by PhanLe1010** [reply]
> Which means if I have 300 volumes they have to cause about 360 kb/s which is totally fine for this amount of volumes. But here is another question for this kind of scale. I know longhorn is using iSCSI. Can it be a bottleneck as I know it's using TCP protocol and even pod and volume located on the same node it's not directly mounted to the pod but all is going through TCP connection?

The flow would be `kernel -> iSCSI initiator -> tgt Linux target -> engine -> replicas`. Only the path from engine -> replicas is TCP connection. `iSCSI initiator -> tgt Linux target` and `tgt Linux target-> engine` paths are unix socket domain connections

In our experience, the bottleneck is usually the TCP connection from engine->replica on different nodes or CPU/RAM limit of the nodes. We don't normally see the bottleneck being the TCP connection from engine -> replica on the same node

That being said, if you don't need replication and always want the pod and volume one the same nodes, you can you use data locality `strict-local` mode. This will make the engine-replica path uses unix socket domain https://longhorn.io/docs/1.6.1/high-availability/data-locality/#:~:text=strict%2Dlocal%3A%20This%20option%20enforces%20Longhorn%20keep%20the%20only%20one%20replica%20on%20the%20same%20node%20as%20the%20attached%20volume%2C%20and%20therefore%2C%20it%20offers%20higher%20IOPS%20and%20lower%20latency%20performance

**Comment by serhiynovos** [comment]
@ejweber  BTW just noticed that on worker nodes where longhorn has replica I see constant write rate about 600kb/s. iostat and iotop show that writing is make mostly by longhorn processes to to the disk and mount point which is scheduled for volumes and write is about 20 - 30 kb for each pvc. As I have now about 30 replicas on this pod I suppose it's expected behavior and write rate.
